widgetry/src/widgets/button.rs

```diff
@@ -271,6 +308,460 @@ impl Btn {
+    /// Has no effect if the button is text-only or image-only.
+    pub fn stack_spacing(mut self, value: f64) -> Self {
+        self.stack_spacing = value;
+        self
+    }
+
+    // Specific UI treatments
+
+    pub fn dropdown(self) -> Self {
+        self.image_path("system/assets/tools/arrow_drop_down.svg")
+            .image_dims(12.0)
+            .stack_spacing(12.0)
+            .label_first()
+    }
+
+    /// Shorthand method to build a Button wrapped in a Widget
+    pub fn build_widget(&self, ctx: &EventCtx, action: &str) -> Widget {
+        Widget::new(Box::new(self.build(ctx, action))).named(action)
+    }
+
+    /// Shorthand method to build a widget whose action is derived from the label's text.
+    // Does `def` stand for anything meaningful? Is there a better short name?
+    pub fn build_def(&self, ctx: &EventCtx) -> Widget {
+        let action = self
+            .default_style
+            .label
+            .as_ref()
+            .and_then(|label| label.text)
+            .expect("Must set `label_text` before calling build_def");
+
+        self.build_widget(ctx, action)
+    }
+
+    pub fn build(&self, ctx: &EventCtx, action: &str) -> Button {
+        let normal = self.batch(ctx, ButtonState::Default);
+        let hovered = self.batch(ctx, ButtonState::Hover);
+        let disabled = self.batch(ctx, ButtonState::Disabled);
+
+        assert!(
+            normal.get_bounds() != geom::Bounds::zero(),
+            "button was empty"
+        );
+        let hitbox = normal.get_bounds().get_rectangle();
+        debug!(
+            "normal.get_bounds().get_rectangle(): {:?} for button: {:?}",
+            hitbox, self
+        );
+        Button::new(
+            ctx,
+            normal,
+            hovered,
+            disabled,
+            self.hotkey.clone(),
+            action,
+            self.tooltip.clone(),
+            hitbox,
+            self.is_disabled,
+        )
+    }
+
+    fn batch(&self, ctx: &EventCtx, for_state: ButtonState) -> GeomBatch {
+        let state_style = self.style(for_state);
+        let default_style = &self.default_style;
+
+        let image_batch = state_style
+            .image
+            .as_ref()
+            .or(default_style.image.as_ref())
+            .and_then(|image| {
+                let default = default_style.image.as_ref();
+                let image_path = image.path.or(default.and_then(|d| d.path));
+                if image_path.is_none() {
+                    return None;
+                }
+                let image_path = image_path.unwrap();
+                let (mut svg_batch, svg_bounds) = svg::load_svg(ctx.prerender, image_path);
+                if let Some(color) = image.color.or(default.and_then(|d| d.color)) {
+                    svg_batch = svg_batch.color(RewriteColor::ChangeAll(color));
+                }
+
+                if let Some(image_dims) = image.dims.or(default.and_then(|d| d.dims)) {
```

Comment:
If I'm reading this correctly, we don't need to duplicate the dimensions in code if we just want to use an image in a button as-is. But being able to scale it _does_ seem quite valuable for putting it next to text.

Reply:
right exactly! 

Uniformly sized icon-only buttons are also a sticking point. It was less of an issue when we were highlighting the icon on hover, but now we highlight a padded area *around* the icon, so it's quite obvious when the icon isn't exactly square or is a slightly different size from its neighbor.

<img width="177" alt="Screen Shot 2021-01-17 at 4 44 59 PM" src="https://user-images.githubusercontent.com/217057/104858218-57df3b00-58e3-11eb-8678-2fbd9870afcc.png">